src/Test/StateMachine/Z.hs

```diff
@@ -0,0 +1,189 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Test.StateMachine.Z
+-- Copyright   :  (C) 2017, ATS Advanced Telematic Systems GmbH
+-- License     :  BSD-style (see the file LICENSE)
+--
+-- Maintainer  :  Stevan Andjelkovic <stevan@advancedtelematic.com>
+-- Stability   :  provisional
+-- Portability :  portable
+--
+-- This module contains Z-inspried combinators for working with relations. The
+-- idea is that they can be used to define concise and showable models. This
+-- module is an experiment and will likely change or move to its own package.
+--
+-----------------------------------------------------------------------------
+
+module Test.StateMachine.Z where
+
+import qualified Data.List as List
+
+------------------------------------------------------------------------
+
+-- | Relations.
+type Rel a b = [(a, b)]
+
+-- | (Partial) functions.
+type Fun a b = Rel a b
+
+------------------------------------------------------------------------
+
+empty :: Rel a b
+empty = []
+
+identity :: [a] -> Rel a a
+identity xs = [ (x, x) | x <- xs ]
+
+singleton :: a -> b -> Rel a b
+singleton x y = [(x, y)]
+
+domain :: Rel a b -> [a]
+domain xys = [ x | (x, _) <- xys ]
+
+codomain :: Rel a b -> [b]
+codomain xys = [ y | (_, y) <- xys ]
+
+isTotalRel :: Eq a => Rel a b -> [a] -> Bool
+isTotalRel r xs = domain r == xs
```

Comment:
Can't use `==` since for example
```haskell
isTotalRel (identity [0,1]) [0] == False
```
Which seems wrong to me, I think the spec should be that `xs` is a subset of the `domain r`

Reply:
The idea is that since we can't enumerate all of `a` (we would need a `Bounded a` and `Enum a` constraint to do that), so we ask the user to supply the universe, i.e. `xs`, against which we check totality.

Arguebly order shouldn't matter though, e.g. `isTotalRel (identity [0, 1]) [1, 0]` should be true... I guess I'll introduce a newtype wrapper for `Rel` with a custom equality or perhaps use `Set`... I like the ergonomics of lists though...

Reply:
For the kind of relations you are modelling here order does not matter, I don't really think there is an argument otherwise.

Or just use a different function instead of `==` here
```haskell
relEq :: Eq a => [a] -> [a] -> Bool
xs `relEq` ys = xs `isSubsetOf` ys && ys `isSubsetOf` xs
```

Reply:
Yeah that's a good idea, thanks.